python/src/main/java/org/apache/apex/malhar/python/base/BasePythonExecutionOperator.java

```diff
@@ -0,0 +1,591 @@
+  }
+
+  @Override
+  public void teardown()
+  {
+    super.teardown();
+    apexPythonEngine.stopInterpreter();
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    super.beginWindow(windowId);
+    requestCounterForThisWindow = 0;
+    responseCounterForThisWindow = 0;
+    numNullResponsesPerWindow = 0;
+    currentWindowId = windowId;
+  }
+
+  @Override
+  public void endWindow()
+  {
+    super.endWindow();
+    if (responseCounterForThisWindow < requestCounterForThisWindow) {
+      LOG.info("Detected stragglers and configured flag/state for blocking at end of window is " +
+          blockAtEndOfWindowForStragglers);
+      if (blockAtEndOfWindowForStragglers) {
+        while (responseCounterForThisWindow < requestCounterForThisWindow) {
+          LOG.debug("Trying to emit all stragglers before the next window can be processed");
+          emitStragglers();
+        }
+      }
+    }
+  }
+
+  @Override
+  public Collection<Partition<BasePythonExecutionOperator>> definePartitions(
+      Collection<Partition<BasePythonExecutionOperator>> partitions, PartitioningContext context)
+  {
+    initPartitioner();
+    return partitioner.definePartitions(partitions,context);
+  }
+
+  @Override
+  public void partitioned(Map<Integer, Partition<BasePythonExecutionOperator>> partitions)
+  {
+    initPartitioner();
+    partitioner.partitioned(partitions);
+  }
+
+  @Override
+  public void beforeCheckpoint(long windowId)
+  {
+    accumulatedCommandHistory.clear();
+    accumulatedCommandHistory.addAll(getApexPythonEngine().getCommandHistory());
+    numStarvedReturnsPerCheckpoint = getApexPythonEngine().getNumStarvedReturns();
+  }
+
+  @Override
+  public void checkpointed(long windowId)
+  {
+    getApexPythonEngine().setNumStarvedReturns(0L);
+    numberOfRequestsProcessedPerCheckpoint = 0;
+  }
+
+  @Override
+  public void committed(long windowId)
+  {
+  }
+
+  public void processPostSetUpPythonInstructions(ApexPythonEngine pythonEngine) throws ApexPythonInterpreterException
+  {
+  }
+
+  @Override
+  public void handleIdleTime()
+  {
+    emitStragglers();
+  }
+
+  @Override
+  public Response processStats(BatchedOperatorStats stats)
+  {
+    Response response = new Response();
+    if ( apexPythonEngine == null) {
+      response.repartitionRequired = true;
+    } else {
+      long starvedCount = apexPythonEngine.getNumStarvedReturns();
+      long requestsForCheckpointWindow = getNumberOfRequestsProcessedPerCheckpoint();
+      float starvationPercent = ((requestsForCheckpointWindow - starvedCount) / requestsForCheckpointWindow) * 100;
+      if (starvationPercent > getStarvationPercentBeforeSpawningNewInstance()) {
+        response.repartitionRequired = true;
+      }
+    }
+    return response;
+  }
+
+  public ApexPythonEngine getApexPythonEngine()
+  {
+    return apexPythonEngine;
+  }
+
+  public void setApexPythonEngine(ApexPythonEngine apexPythonEngine)
```

Comment:
how is this supposed to be used (it appears unused)?

Reply:
This was more of a convenient way to inject a custom interpreter implementation if JEP does not meet their use cases. The current default is JEP and hence the overridden approach need did not exist.